samples: create conversation, create analysis, export data to BigQuery

TrucHLe opened the pull request:
Thank you for opening a Pull Request! Before submitting your PR, there are a few things you can do to make sure it goes smoothly:
- [x] Make sure to open an issue as a [bug/issue](https://github.com/googleapis/nodejs-contact-center-insights/issues/new/choose) before writing your code!  That way we can discuss the change, evaluate designs, and agree on the general idea
- [x] Ensure the tests and linter pass
- [x] Code coverage does not decrease (if any source code was changed)
- [ ] Appropriate docs were updated (if necessary)

Fixes #27 🦕
Fixes #31
Fixes #32 
Fixes #33

TrucHLe committed:
samples: create conversation, create analysis, export data to BigQuery

TrucHLe committed:
Lint

TrucHLe committed:
update test var

TrucHLe committed:
Fix export unit test

TrucHLe committed:
Lint

TrucHLe committed:
Merge branch 'main' into create_del_convo

gcf-owl-bot[bot] committed:
🦉 Updates from OwlBot

See https://github.com/googleapis/repo-automation-bots/blob/main/packages/owl-bot/README.md

TrucHLe committed:
Update README

TrucHLe committed:
update README

TrucHLe committed:
add a space in README

JustinBeckwith commented on samples/exportDataToBigquery.js:
```diff
@@ -0,0 +1,69 @@
+      await operation.promise();
+    } catch (e) {
+      if (
+        e.message ===
```
This feels like a strange corner case to make customers handle.  @alexander-fenster is this expected to be a normal thing?  What are the situations where this could happen?

TrucHLe replied:
I can try to give some context. This is a normal and expected behavior from the Contact Center Insights API. Each time an export operation finishes successfully, it'll return an empty response[1]. The response message is empty by design. If the export operation fails, it'll return an error and it won't match this particular message that we're trying to catch.

I'm not sure why the NodeJS Operation library throws an error when it encounters an empty response (please see line 275 of [2] for this particular error message that we're trying to catch). Ideally NodeJS shouldn't throw an error. But since it does, I want to convey in this sample that "if the API returns this particular error message, it means the export operation finishes successfully", and hence I wrote this `try catch` statement. Let me know how I should update the code to better convey this message!

[1] https://cloud.google.com/contact-center/insights/docs/reference/rpc/google.cloud.contactcenterinsights.v1#exportinsightsdataresponse
[2] https://googleapis.github.io/gax-nodejs/classes/Operation.html#source

alexander-fenster replied:
https://github.com/googleapis/gax-nodejs/pull/1087 will fix it. Please, if you see some weirdness that makes you do things like this, escalate to us (@summer-ji-eng or me), it's better to fix it once and forever than let customers copy-paste the workarounds :)

TrucHLe replied:
Ack! Thank you for the quick fix. I tested my PR with your fix and found that the code ran successfully without the weird `try catch` (I also removed this `try catch`). Thanks again for your help!

JustinBeckwith commented on samples/test/createConversation.test.js:
```diff
@@ -0,0 +1,52 @@
+  });
+
+  afterEach(() => {
+    client.deleteConversation({
```
I am guessing this is async, and we need to `await` it?

TrucHLe replied:
This is a synchronous function for the same reason as the one above in createAnalysis.test.js :-)

JustinBeckwith commented on samples/test/createAnalysis.test.js:
```diff
@@ -0,0 +1,65 @@
+  });
+
+  afterEach(() => {
+    client.deleteConversation({
```
Does this return a promise?  Should we be awaiting?

TrucHLe replied:
IIUC getting a promise is only used for async function. I made this a synchronous function instead of async because I want `deleteConversation` to run synchronously right after the test case. When I wrote this as an async function, the test case fails because there's no timing on when `deleteConversation` will run (e.g. sometimes it'll try to delete the resource before the test finishes creating the resource).

JustinBeckwith replied:
I think I see how you got here.  The `client.deleteConversation` method is very much asynchronous.  If you don't `await` it, that doesn't make it synchronous - it just means the process can complete and exit before the operation is complete.  If the after hook is trying to delete the resource before it's created, that means we're missing an `await` somewhere during creation.  

I would suggest making the lambda passed into the `afterEach` method an `async` function, adding the `await` here, and extending the timeout (if needbe) in `samples/package.json`.  Otherwise, I'm worried the cleanup here isn't always going to complete before the process exits.

TrucHLe replied:
Thanks for the suggestion! I updated 2 tests (CreateAnalysis and CreateConversation) to using `async after` but these are failing because of `NOT FOUND` errors. The other test (CreateConversationWithTtl) that's still using `synchronous after` is still passing. The error message from the 2 failing tests are `NOT FOUND`, saying there was no such resource to delete. But I queried these resources manually in my test project and found that they exist. I'm guessing I wrote `async after` wrong and thus the tests call `deleteConversation` before the conversation was created? Let me know if you're able to spot where I got it wrong 😣

JustinBeckwith commented on samples/test/exportDataToBigquery.test.js:
```diff
@@ -0,0 +1,70 @@
+  let bigqueryProjectId;
+
+  before(async () => {
+    projectId = await client.getProjectId();
```
Are there situations where these two projectIds would be different!?

TrucHLe replied:
Yes! The user can create Contact Center Insights resources in one project and export these resources to BigQuery tables in another project that they own. This feature is documented here https://cloud.google.com/contact-center/insights/docs/export#export_data_to_another_project_optional

TrucHLe commented:
I'm guessing this PR is failing the CI/docs check because I updated the READMEs manually. Previously the READMEs were updated by OwlBot (thanks @tmatsuo for adding the tag!) But I deleted some files and so the previous OwlBot changes aren't correct anymore. I don't have permission to add the `owlbot:run` tag so please help me re-add it!

TrucHLe committed:
update README

TrucHLe committed:
use async delete in test

TrucHLe committed:
remove try catch because LRO error is fixed

sofisl committed:
Merge branch 'main' into create_del_convo

gcf-owl-bot[bot] committed:
🦉 Updates from OwlBot

See https://github.com/googleapis/repo-automation-bots/blob/main/packages/owl-bot/README.md

bcoe commented:
@TrucHLe looks good to me, but it appears that the sample tests are failing:

` No conversation found for project 1046198160504 and conversation ID 17455464441172819646`

TrucHLe committed:
fix tests

TrucHLe commented:
> @TrucHLe looks good to me, but it appears that the sample tests are failing:
> 
> ` No conversation found for project 1046198160504 and conversation ID 17455464441172819646`

Oops, I updated a fix. PTAL

gcf-owl-bot[bot] committed:
🦉 Updates from OwlBot

See https://github.com/googleapis/repo-automation-bots/blob/main/packages/owl-bot/README.md

TrucHLe committed:
update package.json

TrucHLe committed:
change method name

TrucHLe committed:
ignore new docs link

TrucHLe committed:
update comments

TrucHLe committed:
Merge branch 'main' into create_del_convo

TrucHLe committed:
Update README
